### ISSUE TITLE
Implement protobuf format reader

### ISSUE THREAD

**Issue by dgregoire** [confluentinc/ksql#278]
According to the [documentation](https://github.com/confluentinc/ksql/blob/0.1.x/docs/syntax-reference.md#syntax-reference) ksql supports JSON and csv as `VALUE_FORMAT`.

It would be really nice to start testing the concept with higher velocity streams of protobuf data, any plans to support this?

**Comment by hjafarpour** [comment]
@dgregoire We will be adding AVRO support soon. Other formats such as protobuf are on the road map but not prioritized yet. You can also add the protobuf support by adding new serde. Pull requests are always welcome! :)

**Comment by rberenguel** [comment]
@miguno @hjafarpour @dgregoire I'd like to take a look into this (for curiosity on the internals of Ksql + using protobufs + refreshing a bit my terribly old and rusty Java abilities), unless some of you are already giving it a shot, of course

**Comment by dgregoire** [comment]
@rberenguel I'm not currently working on this, would be great to see this

**Comment by miguno** [comment]
Thanks @rberenguel.

FYI: We have already begun looking into additional serdes, but are still going back and forth on the desired UX here.  For example, a great advantage (not the only one! :-P) of KSQL is that all you need is SQL to get stuff done.  And ideally we'd like the same UX for serdes -- for example, you should be able to define everything you need to integrate KSQL with Protobuf through SQL statements (such as defining schemas), if that makes sense.

**Comment by dgregoire** [comment]
@miguno taking the exemple from your documentation

```sql
CREATE STREAM pageviews (viewtime BIGINT, user_id VARCHAR, page_id VARCHAR)
  WITH (VALUE_FORMAT = 'JSON',
        KAFKA_TOPIC = 'my-pageviews-topic');
```

for protobuf, the descriptors of marshaled data are referenced by type and tag, for instance, using the same exemple again,

```sql
CREATE STREAM pageviews (viewtime BIGINT=1, user_id VARCHAR=2, page_id VARCHAR=3)
  WITH (VALUE_FORMAT = 'PROTOBUF',
        KAFKA_TOPIC = 'my-pageviews-topic');
```

**Comment by sfrehse** [comment]
I already working on that feature. However, for each topic you need to provide a message file (.proto) specifying the protobuf message. This is required since protobuf does not encode the schema within the message.

```SQL 
CREATE STREAM pageviews (viewtime BIGINT=1, user_id VARCHAR=2, page_id VARCHAR=3)
  WITH (VALUE_FORMAT = 'PROTOBUF', PROTOBUFSCHEMAFILE="some_file.proto"
        KAFKA_TOPIC = 'my-pageviews-topic');
```

**Comment by sfrehse** [comment]
What are you thinking about loading the compiled protobuf classes via created .jar file within the query, e.g.,

```SQL
CREATE STREAM pageviews (viewtime BIGINT=1, user_id VARCHAR=2, page_id VARCHAR=3)
  WITH (VALUE_FORMAT = 'PROTOBUF', PROTOBUFJARFILE="some_jar_file.jar", PROTOBUFMESSAGENAME="com.my.message" KAFKA_TOPIC = 'my-pageviews-topic');
```

Alternatively, compile a .proto file dynamically during runtime via `protoc`.

**Comment by iaincu** [comment]
I think the jar file is practical for production, so would get my vote, as easiest. Getting really optimistic - any thoughts about generating the stream column definitions from the proto since the information is there already?

**Comment by sfrehse** [comment]
Thanks for your comment. Moreover, I guess there some are other issues. Generally, a topic may contain different kind (logical causal) messages as suggest in, e.g. event sourcing. Are there any ideas belongs to that topic? Maybe some kind of selector of the message to be processed is provided in order to select the correct protobuf message type. However, I guess this is a more general question for AVRO and JSON as well.

**Comment by sfrehse** [comment]
I guess the PR https://github.com/confluentinc/schema-registry/pull/680 is positively related to that issue if not required. Specifying a schema within the query can be similarly done as on that PR.

**Comment by Crim** [comment]
So I spent a little time this morning playing around with adding support for ProtocolBuffers.  In the example shown below, the assumption is that the ProtocolBuffer class is on KSQL's classpath. 

Would anyone from the community be open to reviewing my cobbled together POC and help sort it out into a proper contribution?  I've hacked this together against the 4.1.1 branch as I could not get master to build, but it seems like not much has changed in the APIs in master.

```bash
ksql> CREATE STREAM event_stream (tenantId bigint, objectId bigint) WITH (kafka_topic='EventStream', value_format='PROTOBUF', protobuf_class='org.whatever.protobufs.MyProtoBuf$EventStream');

 Message        
----------------
 Stream created 
----------------
ksql> select tenantId, objectId from event_stream limit 10;
1 | 6
1 | 6
1 | 6

```

**Comment by rodesai** [comment]
That's awesome! As a starting point it would be very helpful if you could write up a document with the proposed API/syntax changes. The Apache Kafka KIPs are a good example (https://cwiki.apache.org/confluence/display/KAFKA/KIP-Template). We are working on setting up a space in the repo for these, but aren't quite ready with it yet. In lieu of that you could put your proposal in a google doc.

You can also file a PR with your changes, but you should do so against master. What problem are you having building?

**Comment by Crim** [comment]
Apologies, master would build, but when I'd attempt to fire up ksql I'd get the following error, running against Kafka 1.1.0

`
[2018-06-20 10:09:52,515] ERROR Failed to initialize TopicClient: Unexpected resource type GROUP for resource 1 (io.confluent.ksql.util.KafkaTopicClient:270)
[2018-06-20 10:09:52,516] ERROR Failed to start KSQL (io.confluent.ksql.rest.server.KsqlServerMain:48)
io.confluent.ksql.util.KsqlException: Could not fetch broker information. KSQL cannot initialize
	at io.confluent.ksql.util.KafkaTopicClientImpl.isTopicDeleteEnabled(KafkaTopicClientImpl.java:271)
	at io.confluent.ksql.util.KafkaTopicClientImpl.<init>(KafkaTopicClientImpl.java:76)
	at io.confluent.ksql.util.KafkaTopicClientImpl.<init>(KafkaTopicClientImpl.java:66)
	at io.confluent.ksql.KsqlEngine.<init>(KsqlEngine.java:153)
	at io.confluent.ksql.KsqlEngine.<init>(KsqlEngine.java:120)
	at io.confluent.ksql.rest.server.KsqlRestApplication.buildApplication(KsqlRestApplication.java:299)
	at io.confluent.ksql.rest.server.KsqlServerMain.createExecutable(KsqlServerMain.java:83)
	at io.confluent.ksql.rest.server.KsqlServerMain.main(KsqlServerMain.java:45)
Caused by: org.apache.kafka.common.errors.InvalidRequestException: Unexpected resource type GROUP for resource 1
`

I'll put together some documentation and clean up this PR

**Comment by rodesai** [comment]
Ah I see - that's an issue with the version of the Kafka Admin Client that KSQL is using. There should be a fix coming soon.

**Comment by Crim** [comment]
Tossed together a PR with the changes hacked together against master -^  

There's considerable work that will need to be done before that will be in an acceptable state, but it 'works.'  I'll start on a Google doc tomorrow outlining the proposed changes and functionality and get it linked on the issue/PR.

**Comment by apurvam** [comment]
thanks @Crim!

We have created a space to create design proposals for KSQL. It would be great if you could follow the process documented here, and be the first person to submit a KsqL Improvement Proposal (aka KLIP): https://github.com/confluentinc/ksql/tree/master/design-proposals

Thanks!

**Comment by Crim** [comment]
Created #1472 PR.  I'm unable to add labels to it.

**Comment by abergmeier** [comment]
@apurvam Could you please guide that process a bit. It seems there is no transparency as to which stage the acceptance of KLIP0 is in.

**Comment by mypark** [comment]
@hjafarpour did this get put on the roadmap?  It's been almost a year so wondering where this netted out.

**Comment by matrixise** [comment]
ping

**Comment by FrancescoPessina** [comment]
Any news on this? Especially with the new release of ksqlDB, would be very very interesting add the possibility to manage protobuffers in ksql.
